futures-watch/src/then_stream.rs

```diff
@@ -0,0 +1,74 @@
+use futures::{Poll, Stream};
+
+use {Watch, WatchError};
+
+/// Maps borrowed references to `T` into an `Item`.
+pub trait Then<T> {
+    /// The output type.
+    type Output;
+
+    /// What you get when Map fails.
+    type Error;
+
+    /// Produces a new Output value.
+    fn then(&mut self, t: &T) -> Result<Self::Output, Self::Error>;
```

Comment:
In futures, `then` receives a `Result<T, E>`. I think if we follow that, then the `Error` enum below won't be needed either, as the `Then` impl can decide how to translate a `WatchError`.